File: scripts/solana_action_agent.py
import argparse
import logging
import os
import asyncio
from datetime import date, timedelta
import base58
import numpy as np
import torch
from dotenv import find_dotenv, load_dotenv
from sklearn.preprocessing import MinMaxScaler
from solana.rpc.async_api import AsyncClient
from solana.keypair import Keypair

from solana_helpers import (
    get_balance,
    get_price_from_pyth,
    swap_sol_to_usdc,
    swap_usdc_to_sol,
    TOKENS
)

# Import our custom model
from models.model1_price_training import Model1, fetch_sol_data

load_dotenv(find_dotenv())

# Load environment variables
PRIVATE_KEY = os.environ.get("SOLANA_PRIVATE_KEY")
RPC_URL = os.environ.get("SOLANA_RPC_URL", "https://api.mainnet-beta.solana.com")

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("agent_logger")

def create_client():
    """Create Solana client and wallet"""
    # Create Solana client
    client = AsyncClient(RPC_URL)
    
    # Create wallet from private key
    private_key_bytes = base58.b58decode(PRIVATE_KEY)
    wallet = Keypair.from_secret_key(private_key_bytes)
    
    return client, wallet

def load_trained_model():
    """Load and return the trained model"""
    model = Model1()
    try:
        model.load_state_dict(torch.load('models/sol_price_model.pth'))
        model.eval()
        return model
    except:
        logger.error("Failed to load trained model. Please run model1_price_training.py first.")
        raise

async def execute_trading_strategy():
    """Execute the trading strategy based on the prediction"""
    logger.info("Starting the trading strategy")
    
    # Create client and wallet
    client, wallet = create_client()

    try:
        # Fetch latest data and prepare features
        features, _, _ = fetch_sol_data(start_date=(date.today() - timedelta(days=60)).strftime("%Y-%m-%d"))
        latest_features = features[-1:]  # Get the most recent data point
        
        # Scale features
        feature_scaler = MinMaxScaler(feature_range=(0, 1))
        features_scaled = feature_scaler.fit_transform(features)
        latest_scaled = feature_scaler.transform(latest_features)
        
        # Load and use our trained model
        model = load_trained_model()
        
        # Make prediction
        with torch.no_grad():
            input_tensor = torch.FloatTensor(latest_scaled)
            prediction = model(input_tensor)
            predicted_price = prediction.item()
        
        logger.info(f"Predicted SOL price: ${predicted_price:.2f}")

        # Get current SOL price from Pyth oracle
        current_sol_price = await get_price_from_pyth("SOL/USD")
        logger.info(f"Current SOL price from Pyth: ${current_sol_price:.2f}")

        # Execute trading logic
        if predicted_price > current_sol_price * 1.01:
            # Check USDC balance
            usdc_balance = await get_balance(client, wallet, TOKENS["USDC"])
            logger.info(f"Your USDC balance: {usdc_balance:.2f}")
            
            if usdc_balance > 0.2:
                logger.info("Buying SOL with 0.1 USDC...")
                tx_hash = await swap_usdc_to_sol(client, wallet, 0.1)
                if tx_hash:
                    logger.info(f"Transaction hash: {tx_hash}")
                else:
                    logger.info("Transaction failed")
            else:
                logger.info("Insufficient USDC balance to buy SOL")
            
        elif predicted_price < current_sol_price * 0.99:
            # Check SOL balance
            sol_balance = await get_balance(client, wallet)
            logger.info(f"SOL balance: {sol_balance:.9f}")
            
            if sol_balance > 0:
                logger.info("Selling SOL for USDC...")
                tx_hash = await swap_sol_to_usdc(client, wallet, sol_balance)
                if tx_hash:
                    logger.info(f"Transaction hash: {tx_hash}")
                else:
                    logger.info("Transaction failed")
            else:
                logger.info("No SOL balance to sell")
        else:
            logger.info("No trading action needed at this time.")

        # Print additional technical indicators
        logger.info("\nTechnical Indicators:")
        logger.info(f"Volume Ratio: {latest_features[0][1]:.2f}")
        logger.info(f"Price Momentum: {latest_features[0][2]:.2f}")

    except Exception as e:
        logger.error(f"Error during trading strategy execution: {str(e)}")
    finally:
        await client.close()
    
    logger.info("-------Agent Execution Finished---------")

if __name__ == "__main__":
    asyncio.run(execute_trading_strategy()) 

File: scripts/solanaActionAgent.ts
import { SolanaAgentKit } from 'solana-agent-kit';
import * as ort from 'onnxruntime-web';
import * as dotenv from 'dotenv';
import {
    getBalance,
    getPriceFromPyth,
    swapSolToUsdc,
    swapUsdcToSol,
    TOKENS
} from './solanaHelpers';
import { fetchSolData } from './fetchSolData';

dotenv.config();

// Load environment variables
const PRIVATE_KEY = process.env.SOLANA_PRIVATE_KEY;
const RPC_URL = process.env.SOLANA_RPC_URL || 'https://api.mainnet-beta.solana.com';
const OPENAI_API_KEY = process.env.OPENAI_API_KEY;

// Simple min-max scaling function
function minMaxScale(data: number[][], featureRange: [number, number] = [0, 1]): number[][] {
    const numFeatures = data[0].length;
    const mins: number[] = new Array(numFeatures).fill(Infinity);
    const maxs: number[] = new Array(numFeatures).fill(-Infinity);

    // Find min and max for each feature
    for (const row of data) {
        for (let i = 0; i < numFeatures; i++) {
            mins[i] = Math.min(mins[i], row[i]);
            maxs[i] = Math.max(maxs[i], row[i]);
        }
    }

    // Scale the data
    return data.map(row =>
        row.map((val, i) => {
            const scaled = (val - mins[i]) / (maxs[i] - mins[i]);
            return scaled * (featureRange[1] - featureRange[0]) + featureRange[0];
        })
    );
}

async function createAgent(): Promise<SolanaAgentKit> {
    if (!PRIVATE_KEY || !OPENAI_API_KEY) {
        throw new Error('Missing required environment variables: SOLANA_PRIVATE_KEY or OPENAI_API_KEY');
    }

    // Create agent with OpenAI API key
    const agent = new SolanaAgentKit(
        PRIVATE_KEY,
        RPC_URL,
        OPENAI_API_KEY
    );

    return agent;
}

async function loadModel(): Promise<ort.InferenceSession> {
    try {
        // Create inference session
        const session = await ort.InferenceSession.create('models/sol_price_model.onnx');
        return session;
    } catch (error) {
        console.error('Failed to load ONNX model:', error);
        throw error;
    }
}

async function executeTradingStrategy() {
    console.log('Starting the trading strategy');

    try {
        // Create agent
        const agent = await createAgent();

        // Fetch latest data and prepare features
        const { features, latestFeatures } = await fetchSolData();

        // Scale features
        const allFeatures = [...features, ...latestFeatures];
        const scaledFeatures = minMaxScale(allFeatures);
        const latestScaled = [scaledFeatures[scaledFeatures.length - 1]];

        // Load model
        const session = await loadModel();

        // Create input tensor
        const inputTensor = new ort.Tensor('float32', new Float32Array(latestScaled.flat()), [1, latestScaled[0].length]);

        // Run inference
        const outputMap = await session.run({ input: inputTensor });
        const predictedPrice = outputMap.output.data[0] as number;

        console.log(`Predicted SOL price: $${predictedPrice.toFixed(2)}`);

        // Get current SOL price from Pyth oracle
        const currentSolPrice = await getPriceFromPyth(agent, 'SOL/USD');
        console.log(`Current SOL price from Pyth: $${currentSolPrice.toFixed(2)}`);

        // Execute trading logic
        if (predictedPrice > currentSolPrice * 1.01) {
            // Check USDC balance
            const usdcBalance = await getBalance(agent, TOKENS.USDC);
            console.log(`Your USDC balance: ${usdcBalance?.toFixed(2)}`);

            if (usdcBalance && usdcBalance > 0.2) {
                console.log('Buying SOL with 0.1 USDC...');
                const txHash = await swapUsdcToSol(agent, 0.1);
                if (txHash) {
                    console.log(`Transaction hash: ${txHash}`);
                } else {
                    console.log('Transaction failed');
                }
            } else {
                console.log('Insufficient USDC balance to buy SOL');
            }
        } else if (predictedPrice < currentSolPrice * 0.99) {
            // Check SOL balance
            const solBalance = await getBalance(agent, TOKENS.SOL);
            console.log(`SOL balance: ${solBalance?.toFixed(9)}`);

            if (solBalance && solBalance > 0) {
                console.log('Selling SOL for USDC...');
                const txHash = await swapSolToUsdc(agent, solBalance);
                if (txHash) {
                    console.log(`Transaction hash: ${txHash}`);
                } else {
                    console.log('Transaction failed');
                }
            } else {
                console.log('No SOL balance to sell');
            }
        } else {
            console.log('No trading action needed at this time.');
        }

        // Print additional technical indicators
        console.log('\nTechnical Indicators:');
        console.log(`Volume Ratio: ${latestFeatures[0][1].toFixed(2)}`);
        console.log(`Price Momentum: ${latestFeatures[0][2].toFixed(2)}`);

    } catch (error) {
        console.error('Error during trading strategy execution:', error);
    }

    console.log('-------Agent Execution Finished---------');
}

// Run the trading strategy
executeTradingStrategy().catch(console.error); 